hoomd/hpmc/shape_move.py

```diff
@@ -0,0 +1,322 @@
+# Copyright (c) 2009-2022 The Regents of the University of Michigan.
+# Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+import hoomd
+from hoomd.operation import _HOOMDBaseObject
+from . import _hpmc
+from hoomd.hpmc import integrate
+from hoomd.data.parameterdicts import ParameterDict
+from hoomd.logging import log
+import numpy
+
+
+class ShapeMove(_HOOMDBaseObject):
+    """Base class for all shape moves.
+
+    A shape move is used as an argument to hoomd.hpmc.update.Shape to specify
+    how to alter shape definitions
+
+    Note:
+        This class should not be instantiated by users. The class can be used
+        for `isinstance` or `issubclass` checks.
+    """
+
+    def _attach(self):
+        self._apply_param_dict()
+        self._apply_typeparam_dict(self._cpp_obj, self._simulation)
+
+
+class Callback(_HOOMDBaseObject):
+    """Base class for callbacks used in Python shape moves.
+
+    Note:
+        This class should not be instantiated by users. User-defined callbacks
+        should inherit from this class, defining a __call__ method that takes
+        a list of floats as an input and returns a shape definition
+
+    Examples::
+
+        class ExampleCallback(hoomd.hpmc.shape_move.Callback):
+            def __call__(self, params):
+                # do something with params and define verts
+                return hoomd.hpmc._hpmc.PolyhedronVertices(verts)
+    """
+
+    def __init__(self):
+        pass
+
+
+class Constant(ShapeMove):
+    """Apply a transition to a specified shape, changing a particle shape by
+    the same way every time the updater is called.
+
+    Note:
+        This is useful for calculating a specific transition probability and
+        derived thermodynamic quantities.
+
+    Args:
+        shape_params (dict): Arguments defining the shape to transition to
+
+    Examples::
+
+        mc = hoomd.hpmc.integrate.ConvexPolyhedron(23456)
+        tetrahedron_verts = [(1, 1, 1), (-1, -1, 1),
+                             (1, -1, -1), (-1, 1, -1)]
+        mc.shape["A"] = dict(vertices=tetrahedron_verts)
+        cube_verts = [(1, 1, 1), (1, 1, -1), (1, -1, 1), (-1, 1, 1),
+                      (1, -1, -1), (-1, 1, -1), (-1, -1, 1), (-1, -1, -1)])
+        constant_move = hoomd.hpmc.shape_move.Constant(shape_params=cube_verts)
+
+    Attributes:
+
+        shape_params (dict): Arguments defining the shape to transition to
+
+    See Also:
+        hoomd.hpmc.integrate for required shape parameters.
+    """
+
+    def __init__(self, shape_params):
+        self._param_dict.update(ParameterDict(shape_params=dict(shape_params)))
```

Comment:
This implicitly lets the user define the schema for the `shape_params` object.

Reply:
I agree, but I didn't find an easy way to validate the schema since the integrator is not needed for construction. We could validate the schema upon attachment and error out if the schema do not matches that of the integrator. What would you suggest?

Reply:
Same applies for the Elastic's `reference` shape. Both attributes are actually `TypeParameters` and I'll fix them once we decide how to address the comment above.

Reply:
Then we would require a shape class to the constructor of this class in the default constructor.

Reply:
Yes, I would appreciate some help with this.